File: js-pr/56/javascript/code-quality-readability-maintainability/json_flattener_47.js
// {fact rule=code-quality-readability-maintainability@v1.0 defects=0}
function flattenNestedJson(data, parentKey = '', separator = '_') {
    const flattenedItems = [];
    
    if (typeof data === 'object' && data !== null && !Array.isArray(data)) {
        Object.entries(data).forEach(([key, value]) => {
            const newKey = parentKey ? `${parentKey}${separator}${key}` : key;
            flattenedItems.push(...flattenValue(value, newKey, separator));
        });
    }
    
    return Object.fromEntries(flattenedItems);
}

function flattenValue(value, key, separator) {
    if (typeof value === 'object' && value !== null && !Array.isArray(value)) {
        return Object.entries(flattenNestedJson(value, key, separator));
    } else if (Array.isArray(value)) {
        return flattenArray(value, key, separator);
    } else {
        return [[key, value]];
    }
}

function flattenArray(items, parentKey, separator) {
    const flattenedItems = [];
    
    items.forEach((item, index) => {
        const indexedKey = `${parentKey}${separator}${index}`;
        flattenedItems.push(...flattenValue(item, indexedKey, separator));
    });
    
    return flattenedItems;
}
// {/fact}

File: python-pr/52/python/code-quality-readability-maintainability/json_flattener_47.py
# {fact rule=code-quality-readability-maintainability@v1.0 defects=0}
def flatten_nested_json(data, parent_key='', separator='_'):
    flattened_items = []
    
    if isinstance(data, dict):
        for key, value in data.items():
            new_key = f"{parent_key}{separator}{key}" if parent_key else key
            flattened_items.extend(_flatten_value(value, new_key, separator))
    
    return dict(flattened_items)

def _flatten_value(value, key, separator):
    if isinstance(value, dict):
        return flatten_nested_json(value, key, separator).items()
    elif isinstance(value, list):
        return _flatten_list(value, key, separator)
    else:
        return [(key, value)]

def _flatten_list(items, parent_key, separator):
    flattened_items = []
    for index, item in enumerate(items):
        indexed_key = f"{parent_key}{separator}{index}"
        flattened_items.extend(_flatten_value(item, indexed_key, separator))
    return flattened_items
# {/fact}